sql/src/main/java/io/crate/planner/operators/HashJoin.java

```diff
@@ -22,42 +22,182 @@
+        this.leftRelation = leftRelation;
+        this.rightRelation = rightRelation;
+        this.joinCondition = joinCondition;
+        this.outputs.addAll(lhs.outputs());
+        this.outputs.addAll(rhs.outputs());
+    }
+
+    public JoinType joinType() {
+        return JoinType.INNER;
     }
 
+    protected Symbol joinCondition() {
+        return joinCondition;
+    }
+
+    public Map<LogicalPlan, SelectSymbol> dependencies() {
+        HashMap<LogicalPlan, SelectSymbol> deps = new HashMap<>(lhs.dependencies().size() + rhs.dependencies().size());
+        deps.putAll(lhs.dependencies());
+        deps.putAll(rhs.dependencies());
+        return deps;
+    }
 
     @Override
-    public ExecutionPlan build(PlannerContext plannerContext, ProjectionBuilder projectionBuilder, int limit, int offset, @Nullable OrderBy order, @Nullable Integer pageSizeHint, Row params, Map<SelectSymbol, Object> subQueryValues) {
-        throw new UnsupportedOperationException("HashJoin not implemented");
+    public ExecutionPlan build(PlannerContext plannerContext,
+                               ProjectionBuilder projectionBuilder,
+                               int limit,
+                               int offset,
+                               @Nullable OrderBy order,
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               Map<SelectSymbol, Object> subQueryValues) {
+
+        ExecutionPlan left = lhs.build(
+            plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
+        ExecutionPlan right = rhs.build(
+            plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
+
+        // Move smaller table to the left side
+        if (rhs.numExpectedRows() < lhs.numExpectedRows()) {
+            ExecutionPlan tmp = left;
+            left = right;
+            right = tmp;
+            AnalyzedRelation tmpRelation = leftRelation;
+            leftRelation = rightRelation;
+            rightRelation = tmpRelation;
+        }
+        ResultDescription leftResultDesc = left.resultDescription();
+        ResultDescription rightResultDesc = right.resultDescription();
+        Collection<String> nlExecutionNodes = ImmutableSet.of(plannerContext.handlerNode());
+
+        MergePhase leftMerge = null;
+        MergePhase rightMerge = null;
+        left.setDistributionInfo(DistributionInfo.DEFAULT_BROADCAST);
+        if (JoinOperations.isNeeded(nlExecutionNodes, leftResultDesc, false)) {
+            leftMerge = JoinOperations.buildForJoin(plannerContext, leftResultDesc, nlExecutionNodes);
+        }
+        if (nlExecutionNodes.size() == 1
+            && nlExecutionNodes.equals(rightResultDesc.nodeIds())
+            && !rightResultDesc.hasRemainingLimitOrOffset()) {
+            // if the left and the right plan are executed on the same single node the mergePhase
```

Comment:
would it be more readable to have the `if` condition extracted into an `executeLeftAndRightPlanOnSameNode` method? (and move the comment inside this method)

Reply:
I don't think it adds any value. If the `if` is extracted then the comment becomes somehow irrelevant on just a boolean condition.